<?xml version="1.0"?>
<launch>
    <arg name="offline" default="false" />

    <arg name="map" value="$(optenv MAP map_output)"/>
    <arg name="map_file" default="$(find mcl_optimal)/maps/$(arg map).pcd"/>
    <arg name="robot_frame" default="base_link"/>
    <arg name="num_particles" default="50" />
    <arg name="lf_obs_std_dev" default="2.0" />
    <arg name="hf_obs_std_dev" default="0.02" />
    
    <group if="$(arg offline)">
    <param name="/use_sim_time" value="false" />
    <node pkg="tf" type="static_transform_publisher" name="base_link_to_rslidar"  args="0.425 0 1.01 0 0 0 /base_link /rslidar 1" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="odom_to_map"  args="0 0 0 0 0 0 odom map 1" />
    <node pkg="mcl_optimal" type="odom_broadcaster_node" name="odom_broadcaster_node" output="screen">
        <param name="robot_frame" value="$(arg robot_frame)" />
    </node>
    <!-- <node pkg="rviz" type="rviz" name="rviz" args="-d $(find mcl_optimal)/config/mcl_optimal.rviz" /> -->
    </group>
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find mcl_optimal)/config/mcl_optimal.rviz" />
    
    
        <!-- <node pkg="pcl_ros" type="pcd_to_pointcloud" name="pcd_to_pointcloud_node" args="$(arg map_file) 0.1 _frame_id:=map" output="screen"/> -->

    <!-- <node pkg="pcl_ros" type="pcd_to_pointcloud" name="pcd_to_pointcloud" args="$(arg map_file)">
        <param name="frame_id" value="map" />
        <remap from="cloud_pcd" to="mapcloud" />
    </node> -->

    <node pkg="mcl_optimal" type="mcl_optimal_node"    name="mcl_optimal_node"    output="screen">
        <param name="map_file" value="$(arg map_file)" />
        <param name="robot_frame" value="$(arg robot_frame)" />
        <param name="num_particles" value="$(arg num_particles)" />
        <param name="lf_obs_std_dev" value="$(arg lf_obs_std_dev)" />
        <param name="hf_obs_std_dev" value="$(arg hf_obs_std_dev)" />
        <param name="init_trans_var_x" value="0.1" />
        <param name="init_trans_var_y" value="0.1" />
        <param name="init_rot_var" value="0.1" />
    </node>

    <!-- <node pkg="normal_estimation" type="approx_normal_estimation_node" name="normal_estimation_node" output="screen">
        <remap from="lidar_points" to="rslidar_points" />
        <param name="max_row_nn_dist" value="0.07" />
        <param name="max_nn_minor_axis_var" value="0.00005" />
        <param name="min_nn_major_axes_var" value="0.0009" />
        <param name="max_pt_to_normal_ang" value="1.3" />
    </node> -->
    <node pkg="normal_estimation" type="normal_estimation_2d" name="normal_estimation_node" output="screen"/>
    <node pkg="normal_estimation" type="normal.py" name="normal_node" output="screen"/>
</launch>
